module/epochs/epoch_lookup.go

```diff
@@ -39,6 +37,33 @@ func (cache *epochRangeCache) latest() epochRange {
 	return cache[2]
 }
 
+// extendLatestEpoch updates the final view of the latest epoch with the final view of the epoch extension.
+// No errors are expected during normal operation.
+func (cache *epochRangeCache) extendLatestEpoch(epochCounter uint64, extension flow.EpochExtension) error {
+	// sanity check: latest epoch should already be cached.
+	if !cache[2].exists() {
+		return fmt.Errorf("sanity check failed: latest epoch does not exist")
+	}
+
+	// sanity check: extensionFinalView should be greater than final view of latest epoch
+	if cache[2].finalView > extension.FinalView {
+		return fmt.Errorf("sanity check failed: latest epoch final view %d greater than extension final view %d", cache[2].finalView, extension.FinalView)
+	}
+
+	// sanity check: epoch extension should have the same epoch counter as the latest epoch
+	if cache[2].counter != epochCounter {
+		return fmt.Errorf("sanity check failed: latest epoch counter %d does not match extension epoch counter %d", cache[2].counter, epochCounter)
+	}
+
+	// duplicate events are no-ops
+	if cache[2].finalView == extension.FinalView {
```

Comment:
I would advice to move de-duplication right after `exists` checks

Reply:
https://github.com/onflow/flow-go/pull/6247/commits/8455caccc14b9837cf164634b426f8f9ca4db874